src/infer_request.cc

```diff
@@ -151,40 +153,46 @@ InferenceRequest::SetState(InferenceRequest::State new_state)
       if (new_state != InferenceRequest::State::STARTED) {
         return generate_error();
       }
-      state_ = new_state;
       IncrementPendingRequestCount();
       break;
     }
     case InferenceRequest::State::STARTED: {
       if (new_state != InferenceRequest::State::EXECUTING) {
         return generate_error();
       }
-      state_ = new_state;
       DecrementPendingRequestCount();
       break;
     }
     case InferenceRequest::State::EXECUTING: {
       if (new_state != InferenceRequest::State::RELEASED) {
         return generate_error();
       }
-      state_ = new_state;
       break;
     }
     case InferenceRequest::State::RELEASED: {
-      // No state transition currently supported after release.
-      return generate_error();
+      if (new_state != InferenceRequest::State::STARTED) {
```

Comment:
I think to reuse the request we'll need to move back to the INITIALIZED state (pending is only incremented in transition from INITIALIZED -> STARTING).

What if we reset state to INITIALIZED as part of PrepareForInference()

@Tabrizian To allow for reusing requests I think we should also reset cancelled state to false in either AsyncInfer() or PrepareForInference() -

Reply:
> What if we reset state to INITIALIZED as part of PrepareForInference()

I like this idea. Would you want to see some default state between construction `new InferenceRequest()` and PrepareForInference() ? Or use initialized for both?

Reply:
> @Tabrizian To allow for reusing requests I think we should also reset cancelled state to false in either AsyncInfer() or PrepareForInference() -

Agreed, I'll make this change.